[Relax][Backend] Fix TVM crashes with default relax pipeline when opt_level=1: InternalError: Check failed: (slot->value_computed) is false

cchung100m opened the pull request:
Hi Commiters,

This PR is trying to fix issues https://github.com/apache/tvm/issues/17876. Any suggestions would be appreciated if you are available.

### Root Cause

`VMShapeLower` crashed when processing `ShapeExpr` containing composite `PrimExpr` that weren't computed yet.

### Solution

~~Modified `VisitExpr_(const ShapeExprNode* op)` in `vm_shape_lower.cc` to:~~

~~1. Mark uncomputed variables as ready for computation~~
~~2. Trigger EmitOutstandingPrimExprCompute() to resolve the dependency chain~~
~~3. Ensure all expressions are computed before calling `MakeSymbolicShapeArg`~~

~~Added test case: `test_composite_shape_expression_fix()` to prevent future occurrences.~~

symbolizing the composite PrimExpr in the pipeline: canonicalization

cchung100m committed:
[#17876]Fix TVM crashes with default relax pipeline when opt_level=1: InternalError: Check failed: (slot->value_computed) is false

cchung100m committed:
Add test case: test_composite_shape_expression

tlopex commented on src/relax/backend/vm/vm_shape_lower.cc:
```diff
@@ -399,6 +399,23 @@ class VMShapeLowerMutator
+        if (it != slot_map_.end() && !it->second->value_computed) {
+          // If it's a variable, mark it as ready for computation
+          if (expr.as<tir::VarNode>()) {
+            it->second->value_computed = true;
```
Marking arbitrary tir::VarNode instances as `value_computed = true` in `VisitExpr_` is incorrect. In VMShapeLower, only Relax ShapeVars (from function parameters/match_cast) or IntImm constants can be safely marked as computed, because only these have runtime values accessible to the VM.
I think correct way is to fix this earlier in the pipeline (shape inference/canonicalization) by symbolizing composite PrimExpr (introduce Relax ShapeVars) or simplifying them before VMShapeLower.

cchung100m committed:
Revert the incorrect solution

cchung100m committed:
new pass: ShapeExprCanonicalizer

cchung100m committed:
Add unit test case for new pass ShapeExprCanonicalizer

cchung100m committed:
Refactor: CanonicalizeShapeExpr

cchung100m committed:
Refactor the canonicalize_shape_expr.cc

cchung100m committed:
Fix lint error UP038

cchung100m committed:
Fix include error

cchung100m commented:
/gemini review

cchung100m committed:
Fix GetOrCreateSymbol crash

cchung100m commented:
/gemini review

cchung100m committed:
Remove the unnecessary handling

cchung100m committed:
Fix builder_->EmitNormalized outside an active binding block would crash

cchung100m commented:
/gemini review

cchung100m committed:
Add test case: test_canonicalize_shape_expr_nested_function

cchung100m commented:
/gemini review

cchung100m committed:
Updte VisitExpr_(const FunctionNode* op)

cchung100m commented:
Hi @tlopex 

Apologies for the delay on this PR. I have just updated the branch with the latest changes and addressed the previous feedback. All CI checks are currently passing. Could you please take another look when you have some time? Let me know if any further adjustments are needed. Thank you so much for your time and patience.
